src/IR/Constraint/Typechecking.hs

```diff
@@ -29,50 +29,64 @@ import Prettyprinter (
   vsep,
  )
 
+import Control.Monad.Writer (
+  tell,
+  when,
+ )
+
 
 typecheckProgram ::
-  I.Program Can.Annotations -> Bool -> Compiler.Pass (I.Program Can.Type, Maybe (Doc ann))
-typecheckProgram pAnn prettyprint = case unsafeTypecheckProgram pAnn prettyprint of
-  Left e -> Compiler.throwError e
-  Right pType -> return pType
+  I.Program Can.Annotations -> Bool -> Compiler.Pass (I.Program Can.Type)
+typecheckProgram pAnn False = do
+  let (result, _) = unsafeTypecheckProgram pAnn False
+  case result of
+    Left e -> Compiler.throwError e
+    Right pType -> return pType
+typecheckProgram pAnn True = do
+  let (result, pp) = unsafeTypecheckProgram pAnn True
+  case result of
+    Left e -> Compiler.dump $ show pp ++ "\n\n" ++ show e
```

Comment:
Ideally, these should be 2 separate `Compiler.Error` cases (`pp` is a `Dump`, and `e` is a `TypeError`). However, the `Compiler.Pass` monad only supports throwing 1 error through the `ExceptT`. @j-hui , is what @jar2333 doing here okay (i.e. putting the type error together with a dump)?

Reply:
It's not the cleanest way to do this, but I'm not opposed to it. What does the output look like, with and without a type error? e.g., what is the output for the following programs:

```
f x = x
```

```
f x = f
```

Reply:
`f x = x`
```
let rec ['ir_t0]
  f : 'ir_t0
given
  exists ['ir_t1, 't4]
    def
      x : 'ir_t1
    in
      local x <: 't4
    ('ir_t1 -> 't4) = 'ir_t0
in
  true

--------------------

f (x: 'ir_t1): 'ir_t0 = x

--------------------

'ir_t0 = ('t4 -> 't4)
'ir_t1 = 't4
'ir_t2 = 't4
'ir_t3 = ('t4 -> 't4)
```

vs

`f x = f`
```
let rec ['ir_t0]
  f : 'ir_t0
given
  exists ['ir_t1, 't4]
    def
      x : 'ir_t1
    in
      local f <: 't4
    ('ir_t1 -> 't4) = 'ir_t0
in
  true

--------------------

f (x: 'ir_t1): 'ir_t0 = f

TypeError (ErrorMsg "Infinite type for variable f, which has the infinite type: Type -> [FlexVar 'ir_t1,Infinite]")
```

I thought it worked fine since currently errors are just printed using the show instance regardless, although of course it can be cleaner and/or open to revision

Reply:
Ok, I think that looks good to me! Maybe add a comment making a note of the fact that you're sort of "misusing" the ExceptT slightly.